README.md
## storm-control ##
This is the Boettiger Lab fork of the Microscope Control Software "Hal", originally written by Hazen Babcock and  the Zhuang Lab. 

[![DOI](https://zenodo.org/badge/15257214.svg)](https://zenodo.org/badge/latestdoi/15257214)

Warning! There have been many changes and this software is now only compatible with Python3 / PyQt5! If you want the end of the line of the (deprecated) Python2 / PyQt4 software then use the python2_pyqt4 branch.

## Installation ##
You will need Python3 and PyQt5 as well as a number of other libraries. Please see the Install.txt file in the hal4000 folder.
To get the storm-analysis dependencies, it is recommended to download the storm-analysis [wheel](https://github.com/ZhuangLab/storm-analysis/releases) and pip install it.  

## General notes ##
1. This software is written primarily in Python with a few C helper libraries.

2. The software is provided "as is" in the hope that others might find it useful. While it is fairly stable and has been developed and used since 2009 in the Zhuang lab, we provide no guarantee that any future changes that are made will maintain backwards compatibility with older versions.

3. We can only provide fairly limited support. You will probably have the most success adapting this software for your purposes if you are reasonably familiar with the Python programming language.


changelog.txt

2021-07-04: Based off current Zhuang lab branch, imported my new modules:
		- sc_hardware/yokogawa  (spinning disk confocal) 
		- sc_hardware/CrestOpitcs  (spinning disk confocal) 
		- sc_hardware/Thorlabs/APTcontroller + module (new stage controllers)
		- sc_hardware/appliedScientificInstrumentation/illuminaStage (new stage controllers)
		- To debug:
			- had to change illuminationChannel to not use startfilm(). apparently I did this earlier too but I've no memory of it.
		- This update fixes two unexpected crash behaviors:
			- When using pgrey lock camera module, Hal crashes when filming (though not when locking). 
			- When parameters are changed (manually or automatically), the focus-lock is unlocked and requires relocking. 
				- Based of testing on scope5, zhuang lab master does not have the 'silent deactivation' of the focus lock created by merging zhuang lab master into Boettiger master. 

2017-07-26: Changed everything in the process of updating for Python3 and
            PyQt5. The entire configuration of a single setup is now
	    specified in a single XML file. See storm_control/hal4000/xml
	    for examples.

2016-02-23: Implement a (basic) parameters editor. This will make it easier
	    for users to change (most) of the parameters.

2016-01-19: Change HAL / Steve to use an absolute pixel size for each 
	    objective rather than a magnification relative to a 100x
	    objective.
	    
2016-01-06: Change the structure of the parameters in preparation for including
	    a parametor editor in HAL. The dot notation will now no longer work.
	    Simplify the parameters that are saved with a movie. These don't
	    need any type information, etc. as they are compared against the
	    default parameters. Move all the less commonly used parameters
	    into the appropriate module.

2015-12-09: Add pixel size metadata to .tif images in a form that ImageJ can
	    understand.

2015-10-09: Fix a recently introduced bug in the Andor EMCCD control software
            that caused it to drop frames and generate an error message in
	    the updateImageWithFrame member of the QCameraWidget class. If
	    you pulled a version in the last month and you use an Andor
	    EMCCD camera an update is recommended.

2015-09-24: Switched to doing all the image manipulation (for image display)
	    in C, falling back to numpy of the C library is not available.
	    
2015-09-21: Implemented camera feeds and cleaned up some of the camera related
	    stuff. At least in theory you can now do things like have multiple
	    displays of the frames from a single camera. You can also display
	    and/or capture average images and images at certain intervals, along
	    with a slice from a larger image. As part of this effort, camera
	    acquisition has been separated from camera display. Hopefully this
	    also makes the camera code a little easier to understand.

	    NOTE! This involves some changes to the layout of the hardware.xml file.

2015-09-09: Implemented control of an Andor sCMOS camera using their SDK3 library.



storm_control/sc_hardware/thorlabs/APTcontroller.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
"""
Connect to a Thorlabs KDC101 controllers using pyAPT
https://github.com/qpit/thorlabs_apt
https://pypi.org/project/thorlabs-apt/#description
https://github.com/qpit/thorlabs_apt/blob/master/thorlabs_apt/core.py

Alistair Boettiger, September 2020

V1.1
Adapted from my PI E873 library


Install notes:
pip install thorlabs-apt   https://pypi.org/project/thorlabs-apt/#description
Downloaded and install the latest version of thorlabs APT 64 bit software from thorlabs
https://www.thorlabs.com/software_pages/ViewSoftwarePage.cfm?Code=Motion_Control
"""


from __future__ import print_function
from copy import deepcopy
import storm_control.sc_library.parameters as params

import thorlabs_apt as apt # The key apt lib


class APTcontroller():

    ## __init__
    #
    # Connect to the PI E873 stage.
    #
    #
    def __init__(self, xStageSN = '27003853', yStageSN = '27003868'): 
        print(['Serial numbers x-stage ' ,xStageSN, ' y-stage ',yStageSN])
    
        self.motorX = apt.Motor(xStageSN)
        self.motorY = apt.Motor(yStageSN)
        
        self.wait = 1 # move commands wait for motion to stop
        self.unit_to_um = 1000.0 #  controller reports in mm    # <--- should become a parameter
        self.um_to_unit = 1.0/self.unit_to_um  # <--- should become a parameter
        
        # # we used the for the memory free KDC controllers
        # self.motorX.set_stage_axis_info(-25,25,1,1.0)
        # self.motorY.set_stage_axis_info(-25,25,1,1.0)
        
        self.motorX.set_stage_axis_info(-250,250,1,1.0)   # <---   should become a parameter
        self.motorY.set_stage_axis_info(-250,250,1,1.0)   # <--- should become a parameter
        # Connect to the stage.
        self.good = 1


    ## getStatus
    #
    # @return True/False if we are actually connected to the stage.
    #
    def getStatus(self):
        return self.good

    ## goAbsolute
    #
    # @param x Stage x position in um.
    # @param y Stage y position in um.
    #
    def goAbsolute(self, x, y):
        if self.good:
            X = x * self.um_to_unit
            Y = y * self.um_to_unit
            rangeX = self.motorX.get_stage_axis_info()
            rangeY = self.motorY.get_stage_axis_info()
            if X > rangeX[0] and X < rangeX[1]:
                print('Moving to X=')
                print(X)
                self.motorX.move_to(X)  # self, value, blocking = False)
            else:
                print('requested move outside max X range!')
                print('requested move');
                print(X)
                print('X range')
                print(rangeX)
            if Y > rangeY[0] and Y < rangeY[1]:
                print('Moving to Y=')
                print(Y)
                self.motorY.move_to(Y)
            else:
                print('requested move outside max Y range!')
                print('requested move');
                print(Y)
                print('Y range')
                print(rangeY)

    ## goRelative
    #
    # @param dx Amount to displace the stage in x in um.
    # @param dy Amount to displace the stage in y in um.
    #
    def goRelative(self, dx, dy):
        if self.good:
            # self.jog(0.0,0.0)
            X =  dx * self.um_to_unit
            Y =  dy * self.um_to_unit
            x0 = self.motorX.position  # query single axis
            y0 = self.motorY.position  # query single axis
            rangeX = self.motorX.get_stage_axis_info()
            rangeY = self.motorY.get_stage_axis_info()
            if x0 + X > rangeX[0] and x0 + X < rangeX[1]:
                print('Moving by X=')
                print(X)
                self.motorX.move_by(X)
            else:
                print('requested move by');
                print(X)
                print('X range')
                print(rangeX)
                print('requested move outside max X range!')
            if y0 + Y > rangeY[0] and y0 + Y < rangeY[1]:
                print('Moving by Y=')
                print(Y)
                self.motorY.move_by(Y)
            else:
                print('requested move by');
                print(Y)
                print('Y range')
                print(rangeY)
                print('requested move outside max Y range!')
            
            
    ## position   https://github.com/qpit/thorlabs_apt/blob/master/thorlabs_apt/core.py
    #
    # @return [stage x (um), stage y (um), stage z (um)]
    #
    def position(self):
        if self.good:
            x0 = self.motorX.position/self.um_to_unit  # query single axis
            y0 = self.motorY.position/self.um_to_unit  # query single axis
            return {"x" : x0,
                "y" : y0}

            
    
    ## jog
    #
    # @param x_speed Speed to jog the stage in x in um/s.  - not clear to me what jog is used for. 
    # @param y_speed Speed to jog the stage in y in um/s.
    #
    def jog(self, x_speed, y_speed):
        pass
        # figure out how to do something here
        # if self.good:
        #     c_xs = c_double(x_speed * self.um_to_unit)
        #     c_ys = c_double(y_speed * self.um_to_unit)
        #     c_zr = c_double(0.0)
        #     tango.LSX_SetDigJoySpeed(self.LSID, c_xs, c_ys, c_zr, c_zr)

    ## joystickOnOff
    #
    # @param on True/False enable/disable the joystick.
    #

    def joystickOnOff(self, on):
        pass
        '''
        if on:
            self.writeline("!joy 2")
        else:
            self.writeline("!joy 0")
        '''
            
    ## lockout
    #
    # Calls joystickOnOff.
    #
    # @param flag True/False.
    #
    def lockout(self, flag):
        self.joystickOnOff(not flag)          

          

    ## setVelocity
    # Not tested yet. I think this should work if we uncomment the last two lines. 
    #
    def setVelocity(self, x_vel, y_vel):
        # get_velocity_parameters(self)  -> (minimum velocity, acceleration, maximum velocity)
        # set_velocity_parameters(self, min_vel, accn, max_vel):
        xVelocityPars = self.motorX.get_velocity_parameters()
        yVelocityPars = self.motorY.get_velocity_parameters()
        print('current velocity parameters, x-stage, y-stage:')
        print(xVelocityPars)
        print(yVelocityPars)
        # self.motorX(xVelocityPars[0],xVelocityPars[1],x_vel)
        # self.motorY(yVelocityPars[0],yVelocityPars[1],y_vel)

    ## shutDown
    #
    # Disconnect from the stage.
    #
    def shutDown(self):
        # Disconnect from the stage
        pass
        
        
    ## zero
    # Not tested yet.  I think this should work if we uncomment the last line. 
    # Set the current position as the new zero position.
    #
    def zero(self):
        if self.good:
            # not sure we need this. Currently, don't reset anything
            xZero = self.motorX.get_move_home_parameters()[2] # the 3rd item is the zero offset
            yZero = self.motorY.get_move_home_parameters()[3] # the 4th item is the zero offset
            print([xZero,yZero])
            # set_move_home_parameters(self, direction, lim_switch, velocity, zero_offset):


storm_control/sc_hardware/pointGrey/pointGreyLockCamera.py
#!/usr/bin/env python
"""
Point Grey camera used in the context of a focus lock.
Hazen 09/19
"""
import numpy
import time
from PyQt5 import QtCore

import storm_control.sc_hardware.utility.af_lock_c as afLC
import storm_control.sc_hardware.pointGrey.spinnaker as spinnaker
import tifffile

# ===== Import fitting libraries. ===========
# Numpy fitter, this should always be available.
import storm_control.sc_hardware.utility.np_lock_peak_finder as npLPF

# Finding/fitting using the storm-analysis project.
saLPF = None
try:
    import storm_control.sc_hardware.utility.sa_lock_peak_finder as saLPF
except ModuleNotFoundError as mnfe:
    print(">> Warning! Storm analysis lock fitting module not found. <<")
    print(mnfe)
    pass

# Finding using the storm-analysis project, fitting using image correlation.
cl2DG = None
try:
    import storm_control.sc_hardware.utility.corr_lock_c2dg as cl2DG
except ModuleNotFoundError as mnfe:
    # Only need one warning about the lack of storm-analysis.
    pass
except OSError as ose:
    print(">> Warning! Correlation lock fitting C library not found. <<")
    print(ose)
    pass
#==============================

class LockCamera(QtCore.QThread):
    """
    This class is used to control a Point Grey (Spinnaker) camera in the 
    context of a focus lock.
    """
    cameraUpdate = QtCore.pyqtSignal(dict)

    def __init__(self, camera_id = None, parameters = None, **kwds):
        super().__init__(**kwds)

        self.cur_offsetx = None
        self.cur_offsety = None        
        self.old_offsetx = None
        self.old_offsety = None
        self.max_offsetx = None
        self.max_offsety = None
        self.n_analyzed = 0
        self.n_dropped = 0
        self.start_time = None

        self.params_mutex = QtCore.QMutex()
        self.running = False
        
        self.zero_dist = parameters.get("zero_dist")

        # Initialize library.
        spinnaker.pySpinInitialize(verbose = False)

        # Get the camera & set some defaults.
        self.camera = spinnaker.getCamera(camera_id)

        # Only Grasshopper has defect correction 
        if self.camera.hasProperty("VideoMode"):
            self.camera.setProperty("VideoMode", parameters.get("video_mode"))
            self.camera.setProperty("pgrDefectPixelCorrectionEnable", False)
        
        # Set pixel format.
        self.camera.setProperty("PixelFormat", "Mono16")  # Set pixel format.  (Mono16)

                
        # We don't want any of these 'features'.
        # self.camera.setProperty("AcquisitionFrameRateAuto", "Off")
        self.camera.setProperty("AcquisitionMode", "Continuous")
        self.camera.setProperty("ExposureAuto", "Off")
        self.camera.setProperty("GainAuto", "Off")        

        if self.camera.hasProperty("pgrExposureCompensationAuto"):
            self.camera.setProperty("pgrExposureCompensationAuto", "Off")

        if self.camera.hasProperty("BlackLevelClampingEnable"):
            self.camera.setProperty("BlackLevelClampingEnable", False)

        if self.camera.hasProperty("SharpnessEnabled"):
            self.camera.setProperty("SharpnessEnabled", False)

        if self.camera.hasProperty("GammaEnabled"):
            self.camera.setProperty("GammaEnabled", False)


        # Verify that we have turned off some of these 'features'.
        for feature in ["pgrDefectPixelCorrectionEnable",
                        "BlackLevelClampingEnable",
                        "SharpnessEnabled",
                        "GammaEnabled"]:
            if self.camera.hasProperty(feature):
                assert not self.camera.getProperty(feature).getValue()

        # Configure camera to not use triggering.
        #
        self.camera.setProperty("TriggerMode", "Off")

        # Configure acquisition parameters.
        #
        # Note: The order is important here.
        #
        for pname in ["BlackLevel", "Gain", "Height", "Width", "OffsetX", "OffsetY", "AcquisitionFrameRate"]:
            self.camera.setProperty(pname, parameters.get(pname))

        # Use maximum exposure time allowed by desired frame rate.
        #
        # self.camera.setProperty("ExposureTime", self.camera.getProperty("ExposureTime").getMaximum())
        self.camera.setProperty("ExposureTime", 2000.0)  # 20000 50 fps  changed to 500 fps

        # Get current offsets.
        #
        self.cur_offsetx = self.camera.getProperty("OffsetX").getValue()
        self.cur_offsety = self.camera.getProperty("OffsetY").getValue()
        self.old_offsetx = self.cur_offsetx
        self.old_offsety = self.cur_offsety
        
        # Set maximum offsets.
        #
        self.max_offsetx = self.camera.getProperty("OffsetX").getMaximum()
        self.max_offsety = self.camera.getProperty("OffsetY").getMaximum()
        
    def adjustAOI(self, dx, dy):
        tmp_x = self.cur_offsetx + dx
        tmp_y = self.cur_offsety + dy

        tmp_x = max(0, tmp_x)
        tmp_x = min(self.max_offsetx, tmp_x)
                
        tmp_y = max(0, tmp_y)
        tmp_y = min(self.max_offsety, tmp_y)

        #
        # The thread loop will check for cur != old and update the camera values
        # as necessary.
        #
        self.params_mutex.lock()
        self.cur_offsetx = tmp_x
        self.cur_offsety = tmp_y
        self.params_mutex.unlock()    

    def adjustZeroDist(self, inc):
        pass

    def run(self):
        self.camera.startAcquisition()
        self.running = True
        while(self.running):
            [frames, frame_size] = self.camera.getFrames()
            self.analyze(frames, frame_size)

            # Check for AOI change.
            self.params_mutex.lock()
            if (self.old_offsetx != self.cur_offsetx) or (self.old_offsety != self.cur_offsety):
                self.camera.stopAcquisition()
                self.camera.setProperty("OffsetX", self.cur_offsetx)
                self.camera.setProperty("OffsetY", self.cur_offsety)
                self.camera.startAcquisition()
                self.old_offsetx = self.cur_offsetx
                self.old_offsety = self.cur_offsety                
            self.params_mutex.unlock()
            self.msleep(5)
            
        self.camera.stopAcquisition()

    def startCamera(self):
        self.start(QtCore.QThread.NormalPriority)
        self.start_time = time.time()
        
    def stopCamera(self, verbose = True):
        if verbose:
            fps = self.n_analyzed/(time.time() - self.start_time)
            print("    > AF: Analyzed {0:d}, Dropped {1:d}, {2:.3f} FPS".format(self.n_analyzed, self.n_dropped, fps))
            print("    > AF: OffsetX {0:d}, OffsetY {1:d}, ZeroD {2:.2f}".format(self.cur_offsetx, self.cur_offsety, self.zero_dist))

        self.running = False
        self.wait()
        self.camera.shutdown()

class CameraQPD(LockCamera):
    """
    QPD emulation class. The default camera ROI of 200x200 pixels.
    The focus lock is configured so that there are two laser spots on the camera.
    The distance between these spots is fit and the difference between this distance and the
    zero distance is returned as the focus lock offset. The maximum value of the camera
    pixels is returned as the focus lock sum.
    """
    def __init__(self, parameters = None, **kwds):
        kwds["parameters"] = parameters
        super().__init__(**kwds)
        
        self.image = None # will be loaded below
        self.fit_hl = None
        self.fit_hr = None
        
        self.cnt = 0
        self.max_backlog = 20
        self.min_good = parameters.get("min_good")
        self.reps = parameters.get("reps")
        self.sum_scale = parameters.get("sum_scale")
        self.sum_zero = parameters.get("sum_zero")
        
        self.good = numpy.zeros(self.reps, dtype = numpy.bool)
        self.mag = numpy.zeros(self.reps)
        self.x_off1 = numpy.zeros(self.reps)
        self.y_off1 = numpy.zeros(self.reps)
        self.x_off2 = numpy.zeros(self.reps)
        self.y_off2 = numpy.zeros(self.reps)
        self.x_width = parameters.get("Width")
        self.y_width = parameters.get("Height")
        
        self.allow_single_fits = False # parameters.get("allow_single_fits") # False 
        self.sigma = parameters.get("sigma") # 5 
        self.background = parameters.get("background") # background 
        self.fit_size = 1.5*self.sigma # parameters.get("fit_size")*self.sigma # 1.5, relative to sigma 
        
        
        # Some derived parameters
        self.half_x = int(self.x_width/2)
        self.half_y = int(self.y_width/2)
        
        # maybe good things to add 
        '''
        self.reps = parameters.get("reps")
        self.sum_scale = parameters.get("sum_scale")
        self.sum_zero = parameters.get("sum_zero")
        self.good = numpy.zeros(self.reps, dtype = numpy.bool)
        self.mag = numpy.zeros(self.reps)
        self.x_off = numpy.zeros(self.reps)
        self.y_off = numpy.zeros(self.reps)
        '''
        
        
    # def adjustAOI is defined above    
        
    def adjustZeroDist(self, inc):
        self.params_mutex.lock()
        self.zero_dist += 0.1*inc
        self.params_mutex.unlock()
    

    def getImage(self):
        return [self.image, self.x_off1, self.y_off1, self.x_off2, self.y_off2, self.sigma]

    def analyze(self, frames, frame_size):

        # Only keep the last max_backlog frames if we are falling behind.
        lf = len(frames)
        if (lf>self.max_backlog):
            self.n_dropped += lf - self.max_backlog
            frames = frames[-self.max_backlog:]
            
        dist1 = numpy.zeros(self.reps)
        dist2 = numpy.zeros(self.reps)
        for elt in frames:
            self.n_analyzed += 1
            frame = elt.getData().reshape(frame_size)
            # frame = elt.getData().reshape((frame_size[1],frame_size[0]))
                            # Convert current frame to 8 bit image.
            image = numpy.right_shift(frame.astype(numpy.uint16), 4).astype(numpy.uint8)
            if self.fit_hl is None:
                roi_size = int(3.0 * self.sigma)
                self.fit_hl = cl2DG.CorrLockFitter(roi_size = roi_size,
                                                   sigma = self.sigma,
                                                   threshold = 10)
                self.fit_hr = cl2DG.CorrLockFitter(roi_size = roi_size,
                                                   sigma = self.sigma,
                                                   threshold = 10)
            
            # Magnitude calculation.
            self.mag[self.cnt] = numpy.max(frame) - numpy.mean(frame)
            
            
            [x1, y1, status1] = self.fit_hl.findFitPeak(image[:,:self.half_x])  # find and fit the spot in the left half of the image
            # consider swapping x and y
            
            self.x_off1[self.cnt] = x1 - self.half_y  # vertical distance from the center of the "image" 
            self.y_off1[self.cnt] = y1 - self.half_x # horizontal distance from the center of "image"
            dist1[self.cnt] = abs(self.half_x - y1)   # distance from middle
                    
            [x2, y2, status2] = self.fit_hr.findFitPeak(image[:,-self.half_x:])  #  find and fit the spot in the right half of the image
            self.x_off2[self.cnt] = x2 - self.half_y # vertical distance from the center of the "image" 
            self.y_off2[self.cnt] = y2
            dist2[self.cnt] = abs(y2)   # distance from middle
            
            tmpOffset = 2.0*((y1 - self.half_x + y2) - self.zero_dist)
            self.good[self.cnt] = status1 and status2
                

                
            if False:
                print('\n\n------------------------\nDebugging findFitPeak')
                print(f'status1 = {status1}')
                print(f'status2 = {status2}')
                print(f'x1 = {self.x_off1[self.cnt]}')
                print(f'y1 = {self.y_off1[self.cnt]}')
                print(f'x2 = {self.x_off2[self.cnt]}')
                print(f'y2 = {self.y_off2[self.cnt]}')
                print(f'm = {self.mag[self.cnt]}')
                print(f'offset = {tmpOffset}')
                
            if False:
                print('\n\n------------------------\nDebugging findFitPeak')
                print(f'x1 = {x1}')
                print(f'y1 = {y1}')
                print(f'x2 = {x2}')
                print(f'y2 = {y2}')
                print(f'm = {self.mag[self.cnt]}')

            # Check if we have all the samples we need.
            self.cnt += 1
            if (self.cnt == self.reps):
                
                #debugging save image to check how it looks.
                # result: frame and image are already scrambled.
                if False:
                    numpy.save(r'C:\Users\STORM1\Desktop\focus_lock_debugging\image.npy',image)
                    numpy.save(r'C:\Users\STORM1\Desktop\focus_lock_debugging\frame.npy',frame)
                
                qpd_dict = {"is_good" : True,
                            "image" : image,
                            "offset" : 0.0,
                            "sigma" : self.sigma,
                            "sum" : 0.0,
                            "x_off1" : 0.0,
                            "y_off1" : 0.0,
                            "x_off2" : 0.0,
                            "y_off2" : 0.0}
                            
                if (numpy.count_nonzero(self.good) < self.min_good):
                    qpd_dict["is_good"] = False
                    self.cameraUpdate.emit(qpd_dict)
                else:
                    mag = numpy.median(self.mag[self.good])
                    m1 =  numpy.median(dist1[self.good])
                    m2 = numpy.median(dist2[self.good])
                    offset = ((m1 + m2) - self.zero_dist)
                    if False:
                        print(f'measured offset = {offset}')
                    qpd_dict["is_good"] = True  # qpd_dict["image"] = image
                    qpd_dict["offset"] = offset
                    qpd_dict["sigma"] = self.sigma
                    qpd_dict["sum"] = self.sum_scale*mag - self.sum_zero
                    qpd_dict["x_off1"] = numpy.mean(self.x_off1[self.good])
                    qpd_dict["y_off1"] = numpy.mean(self.y_off1[self.good])
                    qpd_dict["x_off2"] = numpy.mean(self.x_off2[self.good])
                    qpd_dict["y_off2"] = numpy.mean(self.y_off2[self.good])
                    
                    self.cameraUpdate.emit(qpd_dict)

                self.cnt = 0


    def getZeroDist(self):
        return self.zero_dist


    def shutDown(self):
        super().shutDown()       
        if self.fit_hl is not None:
            self.fit_hl.cleanup()
            self.fit_hr.cleanup()
       


class AFLockCamera(LockCamera):
    """
    This class works with the auto-focus hardware configuration.
    In this configuration there are two spots that move horizontally
    as the focus changes. The spots are shifted vertically so that
    they don't overlap with each other.
    """
    def __init__(self, parameters = None, **kwds):
        kwds["parameters"] = parameters
        super().__init__(**kwds)

        self.cnt = 0
        self.max_backlog = 20
        self.min_good = parameters.get("min_good")
        self.reps = parameters.get("reps")
        self.sum_scale = parameters.get("sum_scale")
        self.sum_zero = parameters.get("sum_zero")

        self.good = numpy.zeros(self.reps, dtype = numpy.bool)
        self.mag = numpy.zeros(self.reps)
        self.x_off = numpy.zeros(self.reps)
        self.y_off = numpy.zeros(self.reps)

        # Create slices for selecting the appropriate regions from the camera.
        t1 = list(map(int, parameters.get("roi1").split(",")))
        self.roi1 = (slice(t1[0], t1[1]), slice(t1[2], t1[3]))

        t2 = list(map(int, parameters.get("roi2").split(",")))
        self.roi2 = (slice(t2[0], t2[1]), slice(t2[2], t2[3]))

        self.afc = afLC.AFLockC(offset = parameters.get("background"),
                                downsample = parameters.get("downsample"))

        assert (self.reps >= self.min_good), "'reps' must be >= 'min_good'."

    def adjustZeroDist(self, inc):
        self.params_mutex.lock()
        self.zero_dist += 0.001*inc
        self.params_mutex.unlock()

    def analyze(self, frames, frame_size):

        # Only keep the last max_backlog frames if we are falling behind.
        lf = len(frames)
        if (lf>self.max_backlog):
            self.n_dropped += lf - self.max_backlog
            frames = frames[-self.max_backlog:]
            
        for elt in frames:
            self.n_analyzed += 1

            frame = elt.getData().reshape(frame_size)
            image1 = frame[self.roi1]
            image2 = frame[self.roi2]
            [x_off, y_off, success, mag] = self.afc.findOffsetU16NM(image1, image2, verbose = False)

            #self.bg_est[self.cnt] = frame[0,0]
            self.good[self.cnt] = success
            self.mag[self.cnt] = mag
            self.x_off[self.cnt] = x_off
            self.y_off[self.cnt] = y_off

            # Check if we have all the samples we need.
            self.cnt += 1
            if (self.cnt == self.reps):

                # Convert current frame to 8 bit image.
                image = numpy.right_shift(frame, 3).astype(numpy.uint8)
                
                qpd_dict = {"is_good" : True,
                            "image" : image,
                            "offset" : 0.0,
                            "sum" : 0.0,
                            "x_off" : 0.0,
                            "y_off" : 0.0}
                            
                if (numpy.count_nonzero(self.good) < self.min_good):
                    qpd_dict["is_good"] = False
                    self.cameraUpdate.emit(qpd_dict)
                else:
                    mag = numpy.mean(self.mag[self.good])
                    y_off = numpy.mean(self.y_off[self.good]) - self.zero_dist

                    qpd_dict["offset"] = y_off
                    qpd_dict["sum"] = self.sum_scale*mag - self.sum_zero
                    qpd_dict["x_off"] = numpy.mean(self.x_off[self.good])
                    qpd_dict["y_off"] = y_off
                    
                    self.cameraUpdate.emit(qpd_dict)

                self.cnt = 0

        
    def stopCamera(self):
        super().stopCamera()
        self.afc.cleanup()


class SSLockCamera(LockCamera):
    """
    This class works with the standard IR laser focus lock configuration.
    In this configuration there is a single spot that movies horizontally
    as the focus changes.
    """
    def __init__(self, parameters = None, **kwds):
        kwds["parameters"] = parameters
        super().__init__(**kwds)

        self.cnt = 0
        self.max_backlog = 20
        self.min_good = parameters.get("min_good")
        self.offset = parameters.get("offset")
        self.reps = parameters.get("reps")
        self.sum_scale = parameters.get("sum_scale")
        self.sum_zero = parameters.get("sum_zero")

        self.good = numpy.zeros(self.reps, dtype = numpy.bool)
        self.mag = numpy.zeros(self.reps)
        self.x_off = numpy.zeros(self.reps)
        self.y_off = numpy.zeros(self.reps)

        self.lpf = slpf.LockPeakFinder(offset = 0,
                                       sigma = parameters.get("sigma"),
                                       threshold = parameters.get("threshold"))

        assert (self.reps >= self.min_good), "'reps' must be >= 'min_good'."

    def adjustZeroDist(self, inc):
        self.params_mutex.lock()
        self.zero_dist += 0.1*inc
        self.params_mutex.unlock()
        
    def analyze(self, frames, frame_size):

        # Only keep the last max_backlog frames if we are falling behind.
        lf = len(frames)
        if (lf>self.max_backlog):
            self.n_dropped += lf - self.max_backlog
            frames = frames[-self.max_backlog:]
            
        for elt in frames:
            self.n_analyzed += 1

            frame = elt.getData().reshape(frame_size)

            # self.offset is slightly below what the camera reads with no
            # signal. We'll be doing MLE fitting so we can't tolerate
            # negative values in 'frame'.
            frame = frame - self.offset

            # Magnitude calculation.
            mag = numpy.max(frame) - numpy.mean(frame)
            
            # Fit peak X/Y location.
            [x_off, y_off, success] = self.lpf.findFitPeak(frame)

            self.good[self.cnt] = success
            self.mag[self.cnt] = mag
            self.x_off[self.cnt] = x_off
            self.y_off[self.cnt] = y_off

            # Check if we have all the samples we need.
            self.cnt += 1
            if (self.cnt == self.reps):

                # Convert current frame to 8 bit image.
                image = numpy.right_shift(frame.astype(numpy.uint16), 3).astype(numpy.uint8)

                mag = numpy.mean(self.mag)
                qpd_dict = {"is_good" : True,
                            "image" : image,
                            "offset" : 0.0,
                            "sum" : self.sum_scale*mag - self.sum_zero,
                            "x_off" : 0.0,
                            "y_off" : 0.0}
                            
                if (numpy.count_nonzero(self.good) < self.min_good):
                    qpd_dict["is_good"] = False
                    self.cameraUpdate.emit(qpd_dict)
                else:
                    y_off = numpy.mean(self.y_off[self.good]) - self.zero_dist

                    qpd_dict["offset"] = y_off
                    qpd_dict["x_off"] = numpy.mean(self.x_off[self.good])
                    qpd_dict["y_off"] = y_off
                    
                    self.cameraUpdate.emit(qpd_dict)

                self.cnt = 0

        
    def stopCamera(self):
        super().stopCamera()
        self.lpf.cleanup()
        

#
# The MIT License
#
# Copyright (c) 2020 Babcock Lab, Harvard University
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#